frontend/src/app/pages/exercises/exercise/shared/simulation/shared/simulation-event-based-report-editor/simulation-event-based-report-editor.component.ts

```diff
@@ -0,0 +1,163 @@
+    @Input() useHotkeys = false;
+
+    private hotkeyLayer: HotkeyLayer | null = null;
+
+    eventBasedReports: EventBasedReport[];
+
+    reportBehaviorState$!: Observable<ReportBehaviorState>;
+
+    constructor(
+        private readonly exerciseService: ExerciseService,
+        private readonly store: Store<AppState>,
+        private readonly hotkeysService: HotkeysService
+    ) {
+        this.eventBasedReports = StrictObject.entries(eventBasedReportData).map(
+            ([eventId, eventDetails]) => ({
+                eventId,
+                enabledKey: eventDetails.enabledKey,
+                description: eventDetails.description,
+                changeCallback: (isEnabled) =>
+                    this.updateEventBasedReport(eventId, isEnabled),
+                hotkey: new Hotkey(eventDetails.hotkeyKeys, false, () =>
+                    this.toggleEventBasedReport(eventId)
+                ),
+            })
+        );
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        if ('useHotkeys' in changes) {
+            if (this.useHotkeys && !this.hotkeyLayer) {
+                this.hotkeyLayer = this.hotkeysService.createLayer();
+                this.registerHotkeys();
+            } else if (!this.useHotkeys && this.hotkeyLayer) {
+                this.hotkeysService.removeLayer(this.hotkeyLayer);
+                this.hotkeyLayer = null;
+            }
+        }
+
+        if ('simulatedRegionId' in changes || 'reportBehaviorId' in changes) {
+            this.reportBehaviorState$ = this.store.select(
+                createSelectBehaviorState<ReportBehaviorState>(
+                    this.simulatedRegionId,
+                    this.reportBehaviorId
+                )
+            );
+        }
+    }
+
+    ngOnDestroy() {
+        if (this.hotkeyLayer) {
+            this.hotkeysService.removeLayer(this.hotkeyLayer);
+            this.hotkeyLayer = null;
+        }
+    }
+
+    registerHotkeys() {
+        if (!this.hotkeyLayer) return;
+
+        this.eventBasedReports.forEach((eventBasedReport) => {
+            this.hotkeyLayer?.addHotkey(eventBasedReport.hotkey);
+        });
+    }
+
+    updateEventBasedReport(type: EventId, isEnabled: boolean) {
+        this.exerciseService.proposeAction({
+            type: eventBasedReportData[type].actionType,
+            simulatedRegionId: this.simulatedRegionId,
+            behaviorId: this.reportBehaviorId,
+            reportChanges: isEnabled,
+        });
+    }
+
+    toggleEventBasedReport(eventId: EventId) {
+        // ESLint seems to be wrong here. Without the type assertion, we receive an ExerciseSimulationBehaviorState
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+        const reportBehavior = selectStateSnapshot(
+            createSelectBehaviorState(
+                this.simulatedRegionId,
+                this.reportBehaviorId
+            ),
+            this.store
+        ) as ReportBehaviorState;
```

Comment:
Maybe use this instead:

```suggestion
        const reportBehavior = selectStateSnapshot(
            createSelectBehaviorState<ReportBehaviorState>(
                this.simulatedRegionId,
                this.reportBehaviorId
            ),
            this.store
        );
```

This should in theory be done to all calls to `createSelectBehaviorState` as its `B` generic parameter can never be inferred from usage. But this should not be something for this PR, I think.

By the way, `createSelectBehaviorState` has a weird contract: It allows you to pass in any `B extends ExerciseSimulationBehaviorState` but filters for a condition unrelated (at the surface at least) to `B`, simply assuming the resulting element is a `B`.

Reply:
Thanks for the suggested fix! I've added it in 255fcd5fcad817d4332f141bada28823ae743fcd

I understand why you think the contract is weird, but I claim that it makes sense given how I expect the selector to be used: If I got a UUID and know from the context what type of behavior to expect, I wouldn't check the type again, but just assume it (i.e., perform a type cast). So I'd rather keep this shorthand way of doing it

Reply:
I see where you're coming from. I think it's fine, at least for now, this PR is already large enough. Still, I think it's a little weird but a lot of selectors are a bit weird.